### ISSUE TITLE
cache directory and stats go out of sync

### ISSUE THREAD

**Issue by timblechmann** [frerich/clcache#70]
in my configuration, the size of the cache dir is configured to be 200GB, but after some time, the cache directory exceeds the stats.txt file: the file reports something like 160GB usage, but the real disk utilisation is about 1TB.

**Comment by timblechmann** [comment]
one possible reason for the mismatch: the `output.txt` files are not taken into account for computing the cache size (some 100kb per translation unit in my codebase)

**Comment by webmaster128** [comment]
Confirmed: my 400 MB cache takes 950 MB on disk.

Given that `output.txt` is typically 1/50 to 1/1000 the size of the actual object, I think there must be something bigger going wrong.

**Comment by webmaster128** [comment]
Can that happen from running multiple clcache processes? E.g.

1. clcache.py A is called and reads stats.txt with `CacheSize = 1000`
1. clcache.py B is called and reads stats.txt with `CacheSize = 1000`
1. clcache.py A starts the real compiler to compile a new object of 100 bytes
1. clcache.py A writes `CacheSize = 1100` to disk
1. clcache.py B starts the real compiler to compile a new object of 200 bytes
1. clcache.py B writes `CacheSize = 1200` to disk

Now the cache size in stats.txt is lower than the file size on disk.

**Comment by frerich** [comment]
It seems to me that the `Configuration` class is inherently racy in its API: it does use a lock file to avoid that there are concurrent read/write accesses to the file, but since the API is defined in terms of getters and setters, there's always the possibility that the configuration is changed on disk without a second `Configuration` object noticing the change (and thus overwriting it when saving). So right now, the burden of maintaining consistency is on the caller.

This is not even specific to the `CacheSize` field, the same race condition would probably manifest when updating e.g. `CallsForLinking` or `CallsWithoutSourceFile`.

It would be possible to change the `Configuration` API such that it is not defined in terms of getter/setters but rather `update` methods (which atomically change a value). However, that's probably going to be inefficient (and certainly incorrect!) when updating multiple values, because the code would acquire and release the lock twice when updating e.g. two fields.

**Comment by webmaster128** [comment]
I guess the most reliable solution here is to use a SQLite database, where we have one connection per clcache.py process. To this database we can send updates like (pseudo code) `CacheSize += 200` or `CallsForLinking += 1`. These are atomic operations that SQLite is able to serialize. This avoids any locks in clcache.

In this case one would not use an in-memory copy of the values but query the database in the getters, giving us the correct values even if changed by other processes.

SQLite files are not as easy to read and debug as JSON files but there a bunch of handy viewers [e.g. DB Browser for SQLite](http://sqlitebrowser.org/). The [sqlite3 module](https://docs.python.org/2/library/sqlite3.html) is available since Python 2.5

**Comment by frerich** [comment]
I don't think any particular backend would solve this issue. The key improvement in your suggestion is to have a way to express `+=`.

Even with SQlite, we would still have the issue that in

* A reads
* B reads
* A writes
* B writes

The second write of `A` would get overwritten. Though I can imagine that some backends can detect a collision of commits and raise a failure (as opposed to just overwriting data).

I'm not arguing for or against and particular backend here but rather mean to say that the flaw is in `clcache` right now in that it assumes that a value read from disc did not change since it was last read, i.e. only the value in memory is incremented and then written out, but the value in memory may have diverged from what's on disk. If SQlite or any other backend happens to make it easy to get this right, I'm all for it. :-)

**Comment by webmaster128** [comment]
SQlite (like other more powerful databases as well) is made to solve exactly this issue. What happens in an update command like `+=` is

* A locks
* B tries to lock and waits
* A reads
* A writes
* A unlocks
* B locks
* B reads
* B writes
* B unlocks

**Comment by timblechmann** [comment]
increasing the contention, one could use the clcache lock to guard the `stat.txt` file? or re-compute the cached size every N compiler invocations (though with the disk performance on windows, we might want to avoid this)

**Comment by frerich** [comment]
Right, @webmaster128 . Note however that the exact same semantics could be implemented with a plain JSON file. We already have a lock which is shared across all processes.

The problem I'm trying to get at is that reading from and writing to the stats file is fairly far apart (in time) right now. In particular, the compiler is run in between if I remember correctly. This means that the (correct) sequence (as outlined in your comment) of having a single lock under which both a read and a write is performed is correct, but greatly limits concurrency because no two `clcache` processes can access the file at the same time (and that time span is almost the lifetime of the `clcache` process). The same issue would be true with `SQlite` or any other backend, as far as I can see.

I believe a good strategy would be to shorten the time during which `clcache` needs exclusive access, such that it's viable to do both operations under a single lock. Once that's done, implementing the correct sequence of steps is going to be easy -- both with the current ad-hoc backend as well as with SQlite.

**Comment by webmaster128** [comment]
> increasing the contention, one could use the clcache lock to guard the stat.txt file?

This is what happens right now but the lock only holds for short periods, not the entire clcache run. This is good because otherwise you'd have e.g. 1 process running and 3 processes waiting on a 4 core machine. What happens is https://github.com/frerich/clcache/issues/70#issuecomment-207998656

> re-compute the cached size every N compiler invocations (though with the disk performance on windows, we might want to avoid this)

The real disk usage data is available in every clean() call (called whenever a new object is written) here: https://github.com/frerich/clcache/blob/master/clcache.py#L167. `objectInfos` stores a list of all objects including their size. The data is just not used.

**Comment by webmaster128** [comment]
I guess you're right @frerich. Maybe we can think of SQlite as a highly optimized version of the file lock you outlined.

Following the principle of minimal steps, it's a good idea to get rid of in-memory state first. To make locking the stats file more efficient, I think we should have separate lock (i.e. clone of ObjectCacheLock) because we should not lock the stats.txt only because an .obj is written to disk, which is a slow operation.

**Comment by timblechmann** [comment]
the cache statistics could be guarded within the lock to ensure that updates to the stats (read-update-write) are atomic.

btw, speaking about locks, i was wondering: is https://github.com/frerich/clcache/blob/master/clcache.py#L246 really safe? my python-foo is a bit rusty ... since the file handle is opened and written under the lock, though i'm wondering if it should also be flushed to disk (or if the GC could do this at an arbitrary point after the lock goes out of scope)

**Comment by frerich** [comment]
@timblechmann I do not believe it's safe in the sense that it appears to rely on specific behaviour of the the GC. I wouldn't be surprised if it does the right thing with CPython, but that's just a particular implementation. A `with` block seems more appropriate here.

I just did a quick Google query and (as usual) StackOverflow had some interesting stuff on it. The answers to [this question](http://stackoverflow.com/questions/7395542/is-explicitly-closing-files-important) seem to agree with my above assessment. So I'd very much welcome a PR which does away with these implicit `close()` calls. ;-)

**Comment by timblechmann** [comment]
@frerich actually i have prototyped an implementation using `with` blocks and compressing `output.txt` files via `GzipFile` ... yet untested, as i haven't done it on a windows machine ...

**Comment by webmaster128** [comment]
I was wondering about this code as well yesterday. See #72

**Comment by frerich** [comment]
Resolved by merging #77 .
